src/basic/process-util.c

```diff
@@ -275,21 +276,59 @@ int get_process_cmdline(pid_t pid, size_t max_length, bool comm_fallback, char *
 }
 
 void rename_process(const char name[8]) {
+        static char *mm = NULL;
+        static size_t mm_size = 0;
+        size_t l;
+        char *old_mm = NULL;
+        size_t old_mm_size = 0;
+
+        /* This is a like a poor man's setproctitle(). It changes the comm field, argv[0], and also the glibc's
+         * internally used name of the process. For the first one a limit of 16 chars applies; to the second one in
+         * many cases one of 10 (i.e. length of "/sbin/init") — however if we have CAP_SYS_RESOURCES it is unbounded;
+         * to the third one 7 (i.e. the length of "systemd". If you pass a longer string it will likely be
+         * truncated. */
+
         assert(name);
 
-        /* This is a like a poor man's setproctitle(). It changes the
-         * comm field, argv[0], and also the glibc's internally used
-         * name of the process. For the first one a limit of 16 chars
-         * applies, to the second one usually one of 10 (i.e. length
-         * of "/sbin/init"), to the third one one of 7 (i.e. length of
-         * "systemd"). If you pass a longer string it will be
-         * truncated */
+        l = strlen(name);
 
+        /* First step, change the comm field. */
         (void) prctl(PR_SET_NAME, name);
 
+        /* Second step, change glibc's ID of the process name. */
         if (program_invocation_name)
                 strncpy(program_invocation_name, name, strlen(program_invocation_name));
 
+        /* Third step, allocate some memory we can place the new name in. To be usable for PR_SET_MM_ARG_START this
+         * needs to be memory suitable for a stack, hence we allocate it with MAP_GROWSDOWN. */
+        if (mm_size < l+1) {
+                size_t nn_size;
+                char *nn;
+
+                nn_size = PAGE_ALIGN(l+1);
+                nn = mmap(NULL, nn_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_GROWSDOWN, -1, 0);
+                if (nn == MAP_FAILED)
+                        goto fallback;
+
+                old_mm = mm;
+                old_mm_size = mm_size;
+
+                mm = nn;
+                mm_size = nn_size;
+        }
+
+        strncpy(mm, name, mm_size);
```

Comment:
So, I've just run `systemd-under-valgrind`:
```sh
bash-4.3# exec /usr/lib/systemd/systemd-under-valgrind
[   17.180668] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[   17.180668]
[   17.181291] CPU: 0 PID: 1 Comm: memcheck-amd64- Not tainted 4.8.12-300.fc25.x86_64 #1
[   17.181291] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.9.3-1.fc25 04/01/2014
[   17.181291]  0000000000000086 00000000577d7eb1 ffff9acd1ea4be00 ffffffffa13e5e6d
[   17.181291]  ffff9acd1ea45700 ffffffffa1c32c68 ffff9acd1ea4be88 ffffffffa11b625c
[   17.181291]  ffffffff00000010 ffff9acd1ea4be98 ffff9acd1ea4be30 00000000577d7eb1
[   17.181291] Call Trace:
[   17.181291]  [<ffffffffa13e5e6d>] dump_stack+0x63/0x86
[   17.181291]  [<ffffffffa11b625c>] panic+0xe4/0x226
[   17.181291]  [<ffffffffa10a5490>] do_exit+0xb10/0xb10
[   17.181291]  [<ffffffffa1253371>] ? __sb_end_write+0x21/0x30
[   17.181291]  [<ffffffffa10a5517>] do_group_exit+0x47/0xb0
[   17.181291]  [<ffffffffa10a5594>] SyS_exit_group+0x14/0x20
[   17.181291]  [<ffffffffa1803932>] entry_SYSCALL_64_fastpath+0x1a/0xa4
[   17.181291] Kernel Offset: 0x20000000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[   17.181291] ---[ end Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[   17.181291]
```
I've reduced to:
```
==17973==
==17973== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==17973==  Access not within mapped region at address 0x4027000
==17973==    at 0x4C3120A: __strncpy_sse2_unaligned (vg_replace_strmem.c:548)
==17973==    by 0x4ECA639: rename_process (process-util.c:320)
==17973==    by 0x10C593: main (test-process-util.c:362)
==17973==  If you believe this happened as a result of a stack
==17973==  overflow in your program's main thread (unlikely but
==17973==  possible), you can try to increase the size of the
==17973==  main thread stack using the --main-stacksize= flag.
==17973==  The main thread stack size used in this run was 2097733632.
==17973==
```
Do we really need that `MAP_GROWSDOWN`-thing?
(given that we always pass 8-10 bytes to the `rename_process`?